src/systemctl/sysv-compat.c

```diff
@@ -65,38 +65,65 @@ int parse_shutdown_time_spec(const char *t, usec_t *ret) {
 
                 *ret = now(CLOCK_REALTIME) + USEC_PER_MINUTE * u;
         } else {
-                char *e = NULL;
-                long hour, minute;
+                char *hour_s;
+                char *min_s;
+                char *colon_position;
+                unsigned long hour, minute;
                 struct tm tm = {};
                 time_t s;
                 usec_t n;
+                bool relative = false;
+                int r;
 
-                errno = 0;
-                hour = strtol(t, &e, 10);
-                if (errno > 0 || *e != ':' || hour < 0 || hour > 23)
+                if (t[0] == '+') {
+                        relative = true;
+                        t++;
+                }
+
+                colon_position = strchr(t,':');
+                if (colon_position == NULL)
+                        return -EINVAL;
+
+                hour_s = strndup(t, colon_position - t);
+                min_s = strndup(colon_position + 1, 2);
+
+                r = safe_atolu(hour_s, &hour);
+                if (r != 0 || hour > 99999)
                         return -EINVAL;
 
-                minute = strtol(e+1, &e, 10);
-                if (errno > 0 || *e != 0 || minute < 0 || minute > 59)
+                r = safe_atolu(min_s, &minute);
+                if (r != 0 || minute > 59)
                         return -EINVAL;
 
+                free(hour_s);
+                free(min_s);
```

Comment:
please use the gcc cleanup logic, i.e. add `_cleanup_free_` to the variable declaration so that these strings are automatically freed when the scope is left. Otherwse it's too easy to forget freeing everything on all error paths, and that#s in fact what happened here: if the safe_atolu() above fails for some reason and you exit early, you forget to free the strings